File: solutions/beecrowd/1025/1025.py
import sys


def binary_search(marbles, query):
    begin = 0
    end = len(marbles) - 1
    middle = end // 2

    while begin <= end:
        if marbles[middle] < query:
            begin = middle + 1
            middle = (end + begin) // 2
        elif marbles[middle] > query or (
            middle > 0 and marbles[middle - 1] == query
        ):
            end = middle - 1
            middle = (end + begin) // 2
        else:
            return middle + 1

    return -1


case = 1

for line in sys.stdin:

    n, q = map(int, line.strip().split())

    if n == 0 and q == 0:
        break

    marbles = []

    while n:
        marble = int(sys.stdin.readline())
        marbles.append(marble)
        n -= 1

    marbles.sort()

    print(f'CASE# {case}:')

    while q:
        query = int(sys.stdin.readline())

        position = binary_search(marbles, query)

        if position == -1:
            print(f'{query} not found')
        else:
            print(f'{query} found at {position}')

        q -= 1

    case += 1

File: solutions/beecrowd/1025/1025.clj
(ns main
  (:require
   [clojure.string :as str]))

(defn binary-search [value items]
  (loop [begin 0
         end (-> (count items)
                 dec)]
    (let [middle (-> (+ begin end)
                     (quot 2))]
      (if  (> begin end)
        -1
        (cond
          (< (items middle) value)
          (recur (inc middle)
                 end)

          (or (> (items middle) value)
              (and (> middle 0)
                   (== (items (dec middle)) value)))
          (recur begin
                 (dec middle))

          :else
          (inc middle))))))

(defn read-values [total-lines]
  (loop [current-line total-lines
         vals []]
    (if (== current-line 0)
      vals
      (recur (dec current-line)
             (cons (Integer/parseInt (read-line)) vals)))))

(defn process-print [test-case-index marbles queries]
  (printf "CASE# %d:%n" test-case-index)
  (loop [queries queries]
    (let [query (first queries)
          position (binary-search query marbles)]
      (if (>= position 0)
        (printf "%d found at %d%n" query position)
        (printf "%d not found%n" query)))
    (when (seq (rest queries))
      (recur (rest queries)))))

(defn main []
  (loop [line (read-line)
         test-case-index 1]
    (let [[n q] (->> (str/split line #" ")
                     (map #(Integer/parseInt %)))
          marbles (-> (read-values n)
                      sort
                      vec)
          queries (-> (read-values q)
                      reverse)]
      (when (and (not= n 0) (not= q 0))
        (process-print test-case-index marbles queries)
        (recur (read-line) (inc test-case-index))))))

(main)
